### PR TITLE
Add mappings versions to CompatibilityVersions

### PR DESCRIPTION
**Pull request by williamrandolph**
CompatibilityVersions now holds a map of system index names to their mappings versions, alongside the transport version. We also add mapping versions to the "minimum version barrier": if a node has a system index whose version is below the cluster mappings version for that system index, it is not allowed to join the cluster.

### CODE REVIEW

**Comment by williamrandolph** [comment]
I think I'm going to back out https://github.com/elastic/elasticsearch/pull/99307/commits/afb6fa080c22c55100d708b41007e97a0df45eff and put it in a follow-up pull request, to make this more reviewable.

**Comment by williamrandolph** [comment]
@elasticsearchmachine please run elasticsearch-ci/part-2

File: server/src/main/java/org/elasticsearch/cluster/ClusterState.java [hunk 1]
```diff
@@ -787,8 +792,9 @@ public DiscoveryNodes nodes() {
             return nodes;
         }
 
+        // TODO[wrb]: Add Map<String, SystemIndexDescriptor.MappingsVersion> argument
```
Comment (williamrandolph) [inline server/src/main/java/org/elasticsearch/cluster/ClusterState.java:795]: I've got a follow-up branch ready for this work.

**Review by ldematte — COMMENTED** [review]
From my limited knowledge, looks good! Only one question: I see the only place where CompatibilityVersions is built with mapping versions is in Node, and elsewhere there are some empty `Map.of()`, like in ClusterState, JoinRequest, etc.. 
Will there be more work to propagate/set these versions or is this OK and the work is done?

**Comment by williamrandolph** [comment]
@ldematte I'll add comments to clarify where the empty Map is deliberate and where it will be cleaned up.

In test code, I'm using an empty map when system index versions aren't relevant to what's being tested.

**Comment by williamrandolph** [comment]
@elasticsearchmachine please run elasticsearch-ci/bwc

File: server/src/main/java/org/elasticsearch/cluster/version/CompatibilityVersions.java [hunk 2]
```diff
@@ -13,49 +13,107 @@
...
+            .map(CompatibilityVersions::transportVersion)
+            .min(Comparator.naturalOrder())
+            // In practice transportVersions is always nonempty (except in tests) but use a conservative default anyway:
+            .orElse(TransportVersions.MINIMUM_COMPATIBLE);
+
+        Map<String, SystemIndexDescriptor.MappingsVersion> minimumMappingsVersions = new HashMap<>();
+        compatibilityVersions.stream()
+            .flatMap(mv -> mv.systemIndexMappingsVersion().entrySet().stream())
+            .forEach(
+                entry -> minimumMappingsVersions.merge(entry.getKey(), entry.getValue(), (v1, v2) -> v1.version() < v2.version() ? v1 : v2)
```
Comment (thecoop) [inline server/src/main/java/org/elasticsearch/cluster/version/CompatibilityVersions.java:59]: You could make `MappingsVersion` implement `Comparable` using the version number so an ordering is part of the class definition

File: server/src/main/java/org/elasticsearch/cluster/version/CompatibilityVersions.java [hunk 3]
```diff
@@ -13,49 +13,107 @@
...
+                    + "] may not join a cluster with minimum transport version ["
+                    + minimumClusterVersions.transportVersion()
+                    + "]"
+            );
+        }
+
+        Map<String, SystemIndexDescriptor.MappingsVersion> candidateInvalid = new HashMap<>();
+        Map<String, SystemIndexDescriptor.MappingsVersion> existingInvalid = new HashMap<>();
+        for (Map.Entry<String, SystemIndexDescriptor.MappingsVersion> candidates : candidate.systemIndexMappingsVersion().entrySet()) {
+            if (minimumClusterVersions.systemIndexMappingsVersion().containsKey(candidates.getKey())
```
Comment (thecoop) [inline server/src/main/java/org/elasticsearch/cluster/version/CompatibilityVersions.java:81]: Pet peeve: double access on a map

```
var mapping = minimumClusterVersions.systemIndexMappingsVersion().get(candidates.getKey())
if (mapping != null && mapping.version() > candidates.getValue().version()) {
   ...
}
```

File: server/src/test/java/org/elasticsearch/cluster/version/CompatibilityVersionsTests.java [hunk 4]
```diff
@@ -10,33 +10,155 @@
...
+            .map(transportVersion -> new CompatibilityVersions(transportVersion, Map.of()))
+            .toList();
 
-        assertThat(CompatibilityVersions.minimumVersions(versionsMap), equalTo(compatibilityVersions1));
+        // should not throw
+        CompatibilityVersions.ensureVersionsCompatibility(
+            new CompatibilityVersions(TransportVersionUtils.randomVersionBetween(random(), min, TransportVersion.current()), Map.of()),
+            compatibilityVersions
+        );
+        expectThrows(
```
Comment (thecoop) [inline server/src/test/java/org/elasticsearch/cluster/version/CompatibilityVersionsTests.java:115]: We should check the exception message too

File: server/src/test/java/org/elasticsearch/cluster/version/CompatibilityVersionsTests.java [hunk 5]
```diff
@@ -10,33 +10,155 @@
...
+        // should not throw
+        CompatibilityVersions.ensureVersionsCompatibility(
+            new CompatibilityVersions(
+                TransportVersion.current(),
+                Map.of(".system-index", new SystemIndexDescriptor.MappingsVersion(min, -1))
+            ),
+            compatibilityVersions
+        );
+        expectThrows(
+            IllegalStateException.class,
```
Comment (thecoop) [inline server/src/test/java/org/elasticsearch/cluster/version/CompatibilityVersionsTests.java:154]: Check the exception message too

**Review by thecoop — APPROVED** [review]
LGTM with some comments

File: server/src/main/java/org/elasticsearch/cluster/version/CompatibilityVersions.java (hunk 2 above)
Comment (williamrandolph) [reply server/src/main/java/org/elasticsearch/cluster/version/CompatibilityVersions.java:59]: That's a good idea. It'll make comparisons in downstream PRs look a lot cleaner.

**Pull request merged**